code/datums/ai/making_your_ai.md

```diff
@@ -0,0 +1,247 @@
+
+```dm
+/datum/ai_controller/haunted
+    planning_subtrees = list(
+        ///someone is currently holding the item,
+        ///preventing it from attacking!
+        ///resist and end planning.
+        /datum/ai_planning_subtree/item_ghost_resist,
+        ///this does not fire this time around
+        /datum/ai_planning_subtree/item_target_from_aggro_list,
+        ///this does not fire this time around
+        /datum/ai_planning_subtree/item_throw_attack,
+    )
+```
+
+### Subtree Setup
+
+Subtrees also have procs for when the mob first starts using them and when they stop. You can use this to make subtrees "reactive" to events via signals, and this is where we set defaults for blackboards if necessary (we want lists to be empty, not null!)
+
+Example:
+
+```
+/datum/ai_planning_subtree/item_ghost_resist/SetupSubtree(datum/ai_controller/controller)
+	RegisterSignal(controller.pawn, COMSIG_ITEM_EQUIPPED, .proc/on_equip)
+	controller.blackboard[BB_LIKES_EQUIPPER] = FALSE
+	controller.blackboard[BB_ITEM_AGGRO_LIST] = list()
+
+/datum/ai_planning_subtree/item_ghost_resist/ForgetSubtree(datum/ai_controller/controller)
+	UnregisterSignal(controller.pawn, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED))
+```
+
+### Behaviors for subtrees
+
+Finally, we've reached the final stop on this controller rabbit hole: Behaviors! These are what subtrees are planning, and the AI will do THESE from first planned all the way to the end, just like it runs through subtrees.
+
+As before, let's take a look at a basic example of one:
+
+```dm
+/datum/ai_behavior/item_escape_grasp
+
+/datum/ai_behavior/item_escape_grasp/perform(delta_time, datum/ai_controller/controller)
+	. = ..()
+	var/obj/item/item_pawn = controller.pawn
+	var/mob/item_holder = item_pawn.loc
+	if(!istype(item_holder))
+		finish_action(controller, FALSE) //We're no longer beind held. abort abort!!
+	item_pawn.visible_message(span_warning("[item_pawn] slips out of the hands of [item_holder]!"))
+	item_holder.dropItemToGround(item_pawn, TRUE)
+	finish_action(controller, TRUE)
+```
+
+This behavior drops the pawn out of the holder's hands. **It must call finish_action to actually finish it, as the behavior will continually be run until finished.** That's important because if you leave no exit condition it's essentially like a while loop that can't end. In this case, whether the behavior is successful or not, it doesn't matter. In theory though:
+
+```dm
+/datum/ai_behavior/item_escape_grasp/finish_action(datum/ai_controller/controller, succeeded)
+	. = ..() <-- ending code for the action
+    if(succeeded)
+        //whatever code for a successful behavior goes here
+```
+
+We can let the behavior ending in a certain way modify the pawn or controller. Most often this is setting/forgetting blackboards.
+
+The last important thing to know is that behaviors take the keys from subtree planning as arguments. They do not search for the blackboards they need themselves.
+
+BAD:
+
+:::danger
```

Comment:
this stuff only works for hackmd, not github

Reply:
Yeah it's kind of a shame i'll cut it just because people probably will read the md file locally